### PR TITLE
perf: replace String and Array indexOf method calls with includes method call

### PR DESCRIPTION
**Pull request by Havunen**
replace String and Array indexOf method calls with includes method call when return value was used only to check that given item exists

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@typescript-bot perf test this faster

**Comment by jakebailey** [comment]
All research I've done on this has shown that `includes` is actually slightly slower (fraction of a percent) than `indexOf` at best in modern v8. But, it is obviously clearer. I doubt it will matter for this codebase.

**Review by RyanCavanaugh — APPROVED** [review]
I triple-checked this for correctness and it looks right. This seems cleaner

File: src/compiler/core.ts [hunk 1]
```diff
@@ -2486,7 +2486,7 @@ export function tryRemoveSuffix(str: string, suffix: string): string | undefined
 
 /** @internal */
 export function stringContains(str: string, substring: string): boolean {
-    return str.indexOf(substring) !== -1;
+    return str.includes(substring);
```
Comment (jakebailey) [inline src/compiler/core.ts:2489]: Eventually it'd be nice to inline this; it has like 40 other calls via this function.

**Review by jakebailey — APPROVED** [review]
Given no real perf change, seems good to me, though I would not complain if you used this PR to replace `stringContains` with `.includes` everywhere.

**Comment by Havunen** [comment]
> Given no real perf change, seems good to me, though I would not complain if you used this PR to replace stringContains with .includes everywhere.

Done

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -30962,7 +30961,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function isHyphenatedJsxName(name: string | __String) {
-        return stringContains(name as string, "-");
+        return (name as string).includes("-");
```
Comment (jakebailey) [inline src/compiler/checker.ts:30964]:
```suggestion
        return name.includes("-");
```

The type assertion isn't needed anymore; it was likely only here as the function only accepted `string`, but the method will be present on both `string` and `__String`.

**Comment by jakebailey** [comment]
Doing a regex search of `\.indexOf\(.*(>|<|<=|>=|===|!==)` shows that there are a few more of these that could be fixed up too. (Some false positives, and some that are `Array`, but mostly fixable).

Not sure what your final goal is but in any case this seems fine to me now and we can replace more later, of course.

**Comment by jakebailey** [comment]
`\.indexOf\(.*(< 0|=== -1|!== -1|>= 0)` is probably a better regex.

File: src/compiler/checker.ts (hunk 2 above)
Comment (Havunen) [reply src/compiler/checker.ts:30964]: I tried it, but it does not compile:

image

**Comment by Havunen** [comment]
> \.indexOf\(.*(< 0|=== -1|!== -1|>= 0) is probably a better regex.

Ok, I went through it and there are some edge cases where indexOf second parameter is used and that cannot be easily replaced with includes check. 

I also didn't want to touch the test files, maybe its better to go through them separately. 

Anyways, if this change is planned to be merged it would be nice to do it before conflicts emerge

**Comment by jakebailey** [comment]
> Ok, I went through it and there are some edge cases where indexOf second parameter is used and that cannot be easily replaced with includes check.

Hm, `includes` has the same second parameter, so I would have expect it to work. But that's fine; I would have taken this regardless.

File: src/compiler/checker.ts (hunk 2 above)
Comment (jakebailey) [reply src/compiler/checker.ts:30964]: Agh, yes, I had some editor delay. Shame.

**Comment by Havunen** [comment]
oh right, I didn't know that :)

**Review by jakebailey — APPROVED** [review]
I manually reviewed all of the changes again, and also diffed this against eslint with eslint-plugin-unicorn's `prefer-includes` rule, and they all line up. Thanks!

**Pull request merged**